autonat/autonat-v2.md

```diff
@@ -0,0 +1,214 @@
+# AutonatV2: spec
+
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+|-----------------|--------------------------|--------|-----------------|
+| 1A              | Working Draft            | Active | r2, 2023-04-15  |
+
+Authors: [@sukunrt]
+
+Interest Group: [@marten-seemann], [@marcopolo], [@mxinden]
+
+[@sukunrt]: https://github.com/sukunrt
+[@marten-seemann]: https://github.com/marten-seemann
+[@mxinden]: https://github.com/mxinden
+[@marcopolo]: https://github.com/marcopolo
+
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is
+publicly reachable. Knowing its NAT status is essential for the node to be
+well-behaved in the network: A node that's behind a NAT / firewall doesn't need
+to advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+In `autonat v2` client sends a priority ordered list of addresses. On receiving
+this list the server dials the first address on the list that it is capable of
+dialing. As the server dials _exactly_ one address from the list, `autonat v2`
+allows nodes to determine reachability for individual addresses. Using `autonat
+v2` nodes can build an address pipeline where they can test individual addresses
+discovered by different sources like identify, upnp mappings, circuit addresses
+etc for reachability. Having a priority ordered list of addresses provides the
+ability to verify low priority addresses. Implementations can generate low
+priority address guesses and add them to requests for high priority addresses as
+a nice to have. This is especially helpful when introducing a new transport.
+Initially, such a transport will not be widely supported in the network.
+Requests for verifying such addresses can be reused to get information about
+other addresses
+
+Compared to `autonat v1` there are two major differences
+1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
+testing reachability for an individual address
+2. `autonat v2` provides a mechanism for nodes to verify whether the peer
+actually successfully dialled an address.
+
+
+## AutoNAT V2 Protocol
+
+
+![Autonat V2 Interaction](autonat-v2.svg)
+
+
+A node wishing to determine reachability of its adddresses sends a `DialRequest`
+message to a peer on a stream with protocol ID `/libp2p/autonat/2.0.0/dial`.
+Each `DialRequest` is sent on a new stream.
+
+This `DialRequest` message has a list of `Candidate`s. Each item in this list
+contains an address and a fixed64 nonce. The list is ordered in descending order
```

Comment:
Would be nice to have a sentence on why you want a unique nonce per address. I think the reasoning is a bit subtle, so would help implementers understand the point.

Reply:
Great point! 
Now I'm not convinced of this requirement. I think a single nonce per request along with verifying that the transport on the local multiaddress where the nonce was received provides us with the same information. 

Consider the case where we are listening on a quic address and a tcp address. 
We ask the peer to dial [(pub_quic_addr, nonce1), (pub_tcp_addr, nonce2)]. It is a malicious peer which decides to dial the tcp address and tell us that it dialed quic address by sending nonce1. So locally our observation will be. 

(local_tcp_addr, pub_quic_addr, nonce1): Even without the address specific nonce we could have figured out that this observation is invalid. The connection we received was a tcp connection and not a quic one. 

Now consider a successful attempt to fool us. We are listening on a quic address and we have two candidates for a public quic address. So we ask the peer to dial (pub_quic_addr1, nonce1), (pub_quic_addr2, nonce2). The malicious peer dials pub_quic_addr2 but sends us nonce1 and informs us that it dialed pub_quic_addr1 with nonce1. 
Here our observation will be (local_quic_addr, pub_quic_addr1, nonce1). We cannot prevent this attack with multiple nonces. 

So I think single nonce plus recording local multiaddress provides us the the same benefits. This will also simplify the request objects.

Reply:
cc @marten-seemann

Reply:
I originally thought the reason to have multiple nonces is so that, assuming honest peers, you could actually figure out which local interface/ip maps to which public IP address. Take your example `(pub_quic_addr1, nonce1), (pub_quic_addr2, nonce2)`. If the peer is honest, I'll know which local interface/ip maps to which public ip by checking the nonce.

However, rereading the spec:
> Upon completion of the dial attempt, the peer sends a `DialResponse` message to
the initiator node on the `/libp2p/autonat/2.0.0/dial` stream with the *index(0
based)* of the candidate that it attempted to dial and the appropriate
`ResponseStatus`

I realize that the honest peer would tell us which public addr it dialed and we'd know which local interface it was because we have the nonce. But the nonce here can just be for the request, not for each addr.

So I agree, I think we only need the single nonce.